File: windows/WallpaperInfoApp/WallpaperInfoApp/WallpaperNotification.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Runtime.InteropServices;

namespace WallpaperInfoApp {

	public partial class WallpaperNotification : Form {
		public WallpaperNotification() {
			InitializeComponent();
			_isOpen = true;
			preparePathsInMenu();
			RegisterHotKeys();
			var themeLabels = ThemeInfo.getLabels();
			int index = 0;
			foreach (var label in themeLabels) {
				var menuItem = new ToolStripMenuItem();
				menuItem.Checked = false;
				menuItem.Tag = index;
				menuItem.Enabled = true;
				menuItem.Text = label;
				menuItem.Click += themeMenuItemSelected;
				_themeMenu.DropDownItems.Add(menuItem);
				index++;
			}
			updateThemeSelectionInMenu(Theme.default1);
            refreshDummyForm();
		}

		void RegisterHotKeys() {
			_platformHotKey = new PlatformHotKey(this);
			_platformHotKey.RegisterHotKey(Keys.F7, KeyModifiers.None);					// Previous (for AppleKeyboard)
			_platformHotKey.RegisterHotKey(Keys.MediaPreviousTrack, KeyModifiers.None);	// Volume Down
			_platformHotKey.RegisterHotKey(Keys.F8, KeyModifiers.None);					// Play/Pause (for AppleKeyboard)
			_platformHotKey.RegisterHotKey(Keys.MediaPlayPause, KeyModifiers.None);		// Play/Pause
			_platformHotKey.RegisterHotKey(Keys.F9, KeyModifiers.None);					// Next (for AppleKeyboard)
			_platformHotKey.RegisterHotKey(Keys.MediaNextTrack, KeyModifiers.None);		// Next
			_platformHotKey.RegisterHotKey(Keys.F11, KeyModifiers.None);				// Volume Down (for AppleKeyboard)
			_platformHotKey.RegisterHotKey(Keys.F10, KeyModifiers.None);				// Volume Mute (for AppleKeyboard)

			// For some reasons, F12, VolumnUP, VolumeMute cannot be hooked.
			//_platformHotKey.RegisterHotKey(Keys.VolumeDown, KeyModifiers.None);		// Volume Down
			//_platformHotKey.RegisterHotKey(Keys.VolumeMute, KeyModifiers.None);		// Volume Mute
			//_platformHotKey.RegisterHotKey(Keys.VolumeUp, KeyModifiers.None);			// Volume Up
			//_platformHotKey.RegisterHotKey(Keys.F12, KeyModifiers.None);				// Volume Up (for AppleKeyboard)

			_platformHotKey.HotKeyPressed += new EventHandler<HotKeyEventArgs>(HotKeyPressed);
		}

		void HotKeyPressed(object sender, HotKeyEventArgs e) {
			BPUtil.BPLog("Hot Key Pressed = " + e.Key.ToString());
			switch (e.Key) {
				case Keys.F8:
				case Keys.MediaPlayPause:
					WallpaperServiceConnection.broadcastToService(MSG.TOGGLE_PAUSE);
					break;

				case Keys.F9:
				case Keys.MediaNextTrack:
					WallpaperServiceConnection.broadcastToService(MSG.NEXT);
					break;

				case Keys.F7:
				case Keys.MediaPreviousTrack:
					WallpaperServiceConnection.broadcastToService(MSG.PREVIOUS);
					break;

				case Keys.F10:
					WallpaperServiceConnection.broadcastToService(MSG.PREVIOUS_THEME);
					break;

				case Keys.F11:
					WallpaperServiceConnection.broadcastToService(MSG.NEXT_THEME);
					break;

				default:
					break;
			}
		}

		public void buildNotification(WPath wpath, Bitmap thumbnail, Theme theme) {
            lock (_self) {
                //BPUtil.BPLog("WallpaperNotification.buildNotification()");
                if (!_isOpen) {
                    return;
                }
                _previousThumbnail = thumbnail;
                MethodInvoker method = delegate {
                    updatePathsInMenu(wpath);
                    setResumeOrPauseOnMenuItem();
                    setThumbnail(wpath, thumbnail);
					updateThemeSelectionInMenu(theme);
                };

                //  choose right thread to call method.
                if (_trayMenuStrip.InvokeRequired) {
                    BeginInvoke(method);
                } else {
                    method.Invoke();
                }
            }
		}

		// Try to avoid the crash in image on strip menu
		private void setThumbnail(WPath wpath, Bitmap thumbnail) {
            try {
                if (thumbnail != null) {
                    if (_menuThumbnail != null) {
                        _menuThumbnail.Dispose();
                    }
                    _menuThumbnail = new Bitmap(thumbnail);
                    _thumbnailMenuItem.BackgroundImage = _menuThumbnail;
                } else {
                    _thumbnailMenuItem.BackgroundImage = global::WallpaperInfoApp.Properties.Resources.thirdwave;
                }
                this._thumbnailMenuItem.Tag = wpath;
                if (wpath != null) {
                    _thumbnailMenuItem.Enabled = true;
                } else {
                    _thumbnailMenuItem.Enabled = false;
                }

            } catch (Exception e) {
                BPUtil.BPLog(e.ToString());
            }
		}

		private void setResumeOrPauseOnMenuItem() {
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			if (wpsi.getPause()) {
				_pauseMenuItem.Text = "Resume";
			} else {
				_pauseMenuItem.Text = "Pause";
			}
		}

		private void preparePathsInMenu() {
            // TODO Popup menu broken in multi monitor due to different DPI 
            // "Per-Monitor Awareness support" needed after Visual Studio 2019
            // autoscale for menu if exists, should be false but not available
			_wallpaperMenu.DropDownItems.Clear();
			for (int index = _maxImageList - 1; index > 0; index--) {
				var menuItem = new ToolStripMenuItem();
				menuItem.Checked = false;
				if (index < _pathsInMenu.Count) {
					menuItem.Text = _pathsInMenu.getWPath(index).label();
					menuItem.Tag = _pathsInMenu.getWPath(index);
					menuItem.Enabled = true;
					menuItem.Click += imageMenuItemClicked;
					if (_pathsInMenu.getWPath(index).path == _previousPath.path) {
						menuItem.Checked = true;
					}
				} else {
					menuItem.Text = "(Empty)";
					menuItem.Tag = null;
					menuItem.Enabled = false;
				}
				_wallpaperMenu.DropDownItems.Add(menuItem);
			}
		}

		private void updatePathsInMenu(WPath wpath) {
			if (wpath == null) {
				// BPUtil.BPLog("WallpaperNotification.updatePathsInMenu() - Error with nil");
				return;
			}
			int previous_index = _pathsInMenu.FirstIndexOf(_previousPath.path);
			if (previous_index >= 0) {
				// there is no other easy way but iterating whole set
				foreach (ToolStripMenuItem menuItem in _wallpaperMenu.DropDownItems) {
					if (previous_index-- == 0) {
						menuItem.Checked = false;
						break;
					}
				}
			}
			_previousPath = wpath;
			String wpathLabel = wpath.label();
			_notifyIcon.Text = wpathLabel;
			int index = _pathsInMenu.FirstIndexOf(wpath.path);
			if (index >= 0) {
				// there is no other easy way but iterating whole set
				foreach (ToolStripMenuItem menuItem in _wallpaperMenu.DropDownItems) {
					if (index-- == 0) {
						menuItem.Checked = true;
						break;
					}
				}
			} else {
				_pathsInMenu.AddFirst(wpath.path, wpath.exif);
				if (_pathsInMenu.Count > _maxImageList) {
					_pathsInMenu.RemoveAt(_maxImageList);
				}
				var menuItem = new ToolStripMenuItem();
				menuItem.Checked = true;
				menuItem.Text = wpathLabel;
				menuItem.Tag = wpath;
				menuItem.Enabled = true;
				menuItem.Click += imageMenuItemClicked;
				_wallpaperMenu.DropDownItems.Insert(0, menuItem);
				if (_wallpaperMenu.DropDownItems.Count > _maxImageList) {
					_wallpaperMenu.DropDownItems.RemoveAt(_maxImageList);
				}
			}			
		}

		private void imageMenuItemClicked(object sender, EventArgs e) {
			if (sender.GetType() == typeof(ToolStripMenuItem)) {
				ToolStripMenuItem menuItem = (ToolStripMenuItem)sender;
				WPath wpath = (WPath)menuItem.Tag;
				if (wpath != null && BPUtil.fileExists(wpath.path)) {
					if (BPUtil.shiftKeyPressed()) {
						BPUtil.showImageFile(wpath.path);
					} else {
						BPUtil.showImagePreview(wpath.path);
					}
				}
			}
		}

		private void updateThemeSelectionInMenu(Theme theme) {
			if (theme != _previousTheme) {
				int index = 0;
				foreach (ToolStripMenuItem menuItem in _themeMenu.DropDownItems) {
					if (theme.intValue() == index) {
						menuItem.Checked = true;
					} else {
						menuItem.Checked = false;
					}
					index++;
				}
				_previousTheme = theme;
			}
		}
	
		private void themeMenuItemSelected(object sender, EventArgs e) {
			if (sender.GetType() == typeof(ToolStripMenuItem)) {
				ToolStripMenuItem menuItem = (ToolStripMenuItem)sender;
				Theme theme = ThemeMethods.rawValue((int)menuItem.Tag);
				WallpaperServiceConnection.broadcastToService(MSG.SET_THEME, theme.intValue());
			}
		}

		private void showWallpaperInfoOptionWindow(object sender, EventArgs e) {
			WallpaperInfoUI.showOptionWindow();
		}

		private void actionPreviousImage(object sender, EventArgs e) {
			WallpaperServiceConnection.broadcastToService(MSG.PREVIOUS);
		}

		private void actionNextImage(object sender, EventArgs e) {
			WallpaperServiceConnection.broadcastToService(MSG.NEXT);
		}

		private void actionPauseResume(object sender, EventArgs e) {
			WallpaperServiceConnection.broadcastToService(MSG.TOGGLE_PAUSE);
		}

		private void notifyIconDoubleClicked(object sender, MouseEventArgs e) {
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			WPath wpath = wpsi.getcurrentWPath();
			if (BPUtil.shiftKeyPressed()) {
				BPUtil.showImageFile(wpath.path);
			} else {
				BPUtil.showImagePreview(wpath.path);
			}
		}

		private void notifyIconClicked(object sender, MouseEventArgs e) {
            // avoid freeze when click and notify happens at the same time
            lock (_self) {
                MouseEventArgs me = (MouseEventArgs)e;
                if (me.Button == System.Windows.Forms.MouseButtons.Left) {
                    WallpaperInfoUI.showOptionWindow(true);
                } else if (me.Button == System.Windows.Forms.MouseButtons.Right) {
                    _notifyIcon.ContextMenuStrip.Show();
                }
            }
		}

		private void trayMenu_Quit(object sender, EventArgs e) {
			_notifyIcon.Visible = false;
			_notifyIcon.Icon = null;
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			if (wpsi.getWallpaperService().isStarted()) {
				wpsi.getWallpaperService().stopService();
			}
			Application.Exit();
		}

        // It looks WIndows uses DPI for pupup menu from the screeen where the last form is opened.
        // So this function helps popup menu to use right DPI by opening dummy form in screen where tray icon is whenever screen is chnaged.
        public static void refreshDummyForm() {
            if (_dummyForm != null) {
                _dummyForm.Close();
            }
            // dummy form helps to avoid broken popup menu in dual screen
            _dummyForm = new Form();
            _dummyForm.ShowInTaskbar = false;
            _dummyForm.ClientSize = new Size(1, 1);
            System.Drawing.Point screenPoint = new System.Drawing.Point(0, 0);
            _dummyForm.Location = _dummyForm.PointToClient(screenPoint);
            _dummyForm.SendToBack();
            _dummyForm.Show();
        }

		// Just to be safe
		private void WallpaperNotificationClosed(object sender, FormClosedEventArgs e) {
			_isOpen = false;
		}

		private bool _isOpen = false;
        private object _self = new object();

		private WLinkedHashMap<String, String> _pathsInMenu = new WLinkedHashMap<String, String>();
		private WPath _previousPath = new WPath("", "");
		private Bitmap _previousThumbnail = null;

		private const int _maxImageList = 20;

		protected override void WndProc(ref Message m) {
            if (PlatformHotKey.HookWndProc(m)) {
                return;
            }
            // for easy deployment
            if (m.Msg == NativeMethods.WM_QUIT_REQUEST) {
                Application.Exit();
                return;
            } else if (m.Msg == NativeMethods.WM_EXTRA_EXECUTED) {
                WallpaperWidgetProvider.updateLabelWidget("App is already running!");
                WallpaperInfoUI.showOptionWindow();
                return;
            // When tablet mode or window mode is changed.
            } else if (m.Msg == NativeMethods.WM_WININICHANGE) {
                if (Marshal.PtrToStringUni(m.LParam) == "UserInteractionMode") {
                    WallpaperWidgetProvider.refreshWidgets();
                }
            }
			base.WndProc(ref m);
		}

		private Theme _previousTheme = Theme.default2;
		private PlatformHotKey _platformHotKey = null;
		private Bitmap _menuThumbnail = null;
        private static Form _dummyForm = null;
    }
}

File: apps/desktop-ubuntu/WallpaperInfoApp/WallpaperNotification.py
#!/usr/bin/env python3

from collections import OrderedDict
import threading

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
gi.require_version('Gdk', '3.0')
from gi.repository import Gdk
from gi.repository import GLib
gi.require_version('Notify', '0.7')
from gi.repository import Notify
gi.require_version('AppIndicator3', '0.1')

import WallpaperInfoUI
import WallpaperServiceInfo as w
import WallpaperServiceConnection
import PlatformHotKey
import BPUtil
import WPath
import ThemeInfo as t
import AppUtil as a

_maxImageList = 15

class WallpaperNotification:

    def __init__(self, app_id, icon):
        self._lock = myLock()
        builder = Gtk.Builder()
        builder.add_from_file('WallpaperInfoApp.glade')
        self.initWithBuilder(builder)
        # init properties
        self._app_id = app_id
        self._icon = icon
        self._menuThumbnail = None
        self._isOpen = True
        self._pathsInMenu = MyOrderedDict()
        self._previousPath = WPath.WPath("", "")
        self._previousThumbnail = None
        self._previousTheme = t.Theme.all
        self.setThumbnail(None, None)
        # set up optionUI / customconfig window

        self._wallpaperInfoUI = WallpaperInfoUI.WallpaperInfoUI(builder, icon)
        self.connectSignals(builder)
        # initialize menu items
        self.preparePathsInMenu()
        themeLabels = t.getLabels()
        self._themesMenuItem.set_submenu(Gtk.Menu())
        index = 0
        for themelabel in themeLabels:
            menuItem = Gtk.CheckMenuItem(label=themelabel)
            menuItem.connect('activate', self.themeMenuItemSelected)
            menuItem.set_active(False)
            menuItem.name = str(index)
            self._themesMenuItem.get_submenu().append(menuItem)
            index += 1
        self._themesMenuItem.show_all()

        # initialize app indicator
        self.initializeAppIndicator()
        # set up hot key
        self._hotKeys = PlatformHotKey.PlatformHotKey()
        self._hotKeys.beginHotKey()

    def initWithBuilder(self, builder):
        self._mainMenu = builder.get_object('MainMenu')
        self._wallpaperMenuItem = builder.get_object('wallpaperMenuItem')
        self._themesMenuItem = builder.get_object('themesMenuItem')
        self._pauseMenuItem = builder.get_object('pauseMenuItem')
        self._thumbnailMenuItem = builder.get_object('thumbnailMenuItem')
        self._thumbnailMenuItemImage = builder.get_object('thumbnailMenuItemImage')
        self._optionsMenuItem = builder.get_object('optionsMenuItem')

    def connectSignals(self, builder):
        builder.connect_signals({
            "menuItemPreviousImage": self.menuItemPreviousImage,
            "menuItemNextImage": self.menuItemNextImage,
            "menuItemPauseResume": self.menuItemPauseResume,
            "imageMenuItemClicked": self.imageMenuItemClicked,
            "onShowOrHide": self.onShowOrHide,
            "onAbout": self.onAbout,
            "onQuit": self.onQuit,

            "onDirButtonClicked": self._wallpaperInfoUI.onDirButtonClicked,
            "onIntervalScaleValueChanged": self._wallpaperInfoUI.onIntervalScaleValueChanged,
            "onThemeComboBoxChanged": self._wallpaperInfoUI.onThemeComboBoxChanged,
            "thumbnailImageClicked": self._wallpaperInfoUI.thumbnailImageClicked,
            "onStartStopButtonClicked": self._wallpaperInfoUI.onStartStopButtonClicked,
            "onPreviousButtonClicked": self._wallpaperInfoUI.onPreviousButtonClicked,
            "onPauseButtonClicked": self._wallpaperInfoUI.onPauseButtonClicked,
            "onNextButtonClicked": self._wallpaperInfoUI.onNextButtonClicked,
            "customConfigStringSetButtonClicked": self._wallpaperInfoUI.customConfigStringSetButtonClicked,
        })

    def __del__(self):
        self._hotKeys.endHotKey()

    def buildNotification(self, wpath, thumbnail, theme):
        # BPUtil.BPLog("WallpaperNotification.buildNotification()")
        if not self._isOpen:
            return
        self._previousThumbnail = thumbnail
        Gdk.threads_add_idle(GLib.PRIORITY_DEFAULT_IDLE, self.updatePathsInMenu, wpath)
        Gdk.threads_add_idle(GLib.PRIORITY_DEFAULT_IDLE, self.setResumeOrPauseOnMenuItem)
        Gdk.threads_add_idle(GLib.PRIORITY_DEFAULT_IDLE, self.setThumbnail, wpath, thumbnail)
        Gdk.threads_add_idle(GLib.PRIORITY_DEFAULT_IDLE, self.updateThemeSelectionInMenu, theme)

    def setThumbnail(self, wpath, thumbnail):
        if thumbnail != None:
            if self._menuThumbnail != None:
                del self._menuThumbnail
            self._menuThumbnail = thumbnail
            self._thumbnailMenuItemImage.set_from_pixbuf(thumbnail)
            self._thumbnailMenuItem.set_label(wpath.label())
        else:
            # to use set_from_pixbuf: pixbuf = Gtk.Image.new_from_file(self._icon)
            self._thumbnailMenuItem.set_label("Waiting for Image")
            self._thumbnailMenuItemImage.set_from_file(self._icon)
        if (wpath != None):
            self._thumbnailMenuItem.name = wpath.path
            self._thumbnailMenuItem.set_sensitive(True)
        else:
            self._thumbnailMenuItem.name = None
            self._thumbnailMenuItem.set_sensitive(False)

    def setResumeOrPauseOnMenuItem(self):
        wpsi = w.WallpaperServiceInfo()
        if (wpsi.getPause()):
            self._pauseMenuItem.set_label("Resume")
        else:
            self._pauseMenuItem.set_label("Pause")

    def preparePathsInMenu(self):
        self._wallpaperMenuItem.set_submenu(Gtk.Menu())
        for index in range(0,_maxImageList):
            menuItem = Gtk.CheckMenuItem()
            menuItem.set_active(False)
            if index < len(self._pathsInMenu):
                wpath = self._pathsInMenu.items()[index]
                menuItem.set_label(wpath.label())
                menuItem.name = wpath.path
                menuItem.set_sensitive(True)
                menuItem.connect('activate', self.imageMenuItemClicked)
                if wpath.path == self._previousPath.path:
                    menuItem.set_active(True)
            else:
                menuItem.set_label("(Empty)")
                menuItem.name = None
                menuItem.set_sensitive(False)
            self._wallpaperMenuItem.get_submenu().append(menuItem)
        self._wallpaperMenuItem.show_all()

    def updatePathsInMenu(self, wpath):
        if wpath == None:
            #BPUtil.BPLog("WallpaperNotification.updatePathsInMenu - Error with None")
            return
        previous_index = self._pathsInMenu.firstIndexOf(self._previousPath.path)
        if previous_index >= 0:
            # there is no other easy way but iterating whole set
            for menuItem in self._wallpaperMenuItem.get_submenu().get_children():
                if previous_index == 0:
                    menuItem.set_active(False)
                    break
                previous_index -= 1
        self._previousPath = wpath
        wpathLabel = wpath.label()
        index = self._pathsInMenu.firstIndexOf(wpath.path)
        if index >= 0:
            # there is no other easy way but iterating whole set
            for menuItem in self._wallpaperMenuItem.get_submenu().get_children():
                if index == 0:
                    menuItem.set_active(True)
                    break
                index -= 1
        else:
            # value is not used.
            self._pathsInMenu.prepend(wpath.path, wpath.exif)
            if len(self._pathsInMenu) > _maxImageList:
                self._pathsInMenu.popItem()
            menuItem = Gtk.CheckMenuItem(label=wpathLabel)
            menuItem.set_active(True)
            menuItem.set_label(wpathLabel)
            menuItem.set_sensitive(False)
            # TODO activate impl by distinguish between actual click and call
            # menuItem.connect('activate', self.imageMenuItemClicked)
            self._wallpaperMenuItem.get_submenu().insert(menuItem, 0)
            menuitems_len = len(self._wallpaperMenuItem.get_submenu().get_children())
            if menuitems_len > _maxImageList:
                menuitem_to_delete = self._wallpaperMenuItem.get_submenu().get_children()[menuitems_len - 1]
                self._wallpaperMenuItem.get_submenu().remove(menuitem_to_delete)
        self._wallpaperMenuItem.show_all()

    def updateThemeSelectionInMenu(self, theme):
        theme_int = int(theme)
        if theme_int != self._previousTheme:
            self._previousTheme = theme_int
            index = 0
            menuItems = self._themesMenuItem.get_submenu().get_children()
            for menuItem in menuItems:
                # only in GTK, set_active call triggers themeMenuItemSelected
                if theme_int != index:
                    if menuItem.get_active():
                        menuItem.set_active(False)
                        menuItem.set_sensitive(True)
                else:
                    menuItem.set_active(True)
                    menuItem.set_sensitive(False)
                index += 1

    def themeMenuItemSelected(self, *args):
        menuItem = args[0]
        theme_int = int(menuItem.name)
        if theme_int != self._previousTheme and menuItem.get_active():
            WallpaperServiceConnection.broadcastToService(a.MSG.SET_THEME, theme_int)

    def menuItemPreviousImage(self, *args):
        WallpaperServiceConnection.broadcastToService(a.MSG.PREVIOUS)

    def menuItemNextImage(self, *args):
        WallpaperServiceConnection.broadcastToService(a.MSG.NEXT)

    def menuItemPauseResume(self, *args):
        WallpaperServiceConnection.broadcastToService(a.MSG.TOGGLE_PAUSE)

    def imageMenuItemClicked(self, *args):
        menuitem = args[0]
        path = menuitem.name
        if path != None and BPUtil.fileExists(path):
            BPUtil.showImageFile(path)
            #BPUtil.showImagePreview(path)

    def onShowOrHide(self, *args):
        WallpaperInfoUI.showOptionWindow(True)

    def onAbout(self, *args):
        Notify.Notification.new("About WallpaperInfoApp",
            "WallpaperInfoApp 1.0 by Brad Park",
            self._icon).show()

    def onQuit(self, *args):
        wpsi = w.WallpaperServiceInfo()
        wpsi.getWallpaperService().stopService()
        Notify.uninit()
        Gtk.main_quit()

    def initializeAppIndicator(self):
        appIndicatorSupport = True
        try:
            from gi.repository import AppIndicator3
        except:
            appIndicatorSupport = False
        if appIndicatorSupport:
            self.ind = AppIndicator3.Indicator.new(
                self._app_id, self._icon,
                AppIndicator3.IndicatorCategory.APPLICATION_STATUS)
            self.ind.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
            self.ind.set_menu(self._mainMenu)
            self.ind.set_secondary_activate_target(self._optionsMenuItem)
        else:
            self.ind = Gtk.StatusIcon()
            self.ind.set_from_file(self._icon)
            self.ind.connect('popup-menu', self.onPopupMenu)
        Notify.init(self._app_id)

    # not used
    def onPopupMenu(self, icon, button, time):
        self.menu.popup(None, None, Gtk.StatusIcon.position_menu, icon,
                        button, time)

class myLock():
    def __init__(self):
        self._lock = threading.Lock()
    def acquire(self, tag = ""):
        #print(tag, "lock acquire")
        self._lock.acquire()
    def release(self, tag = ""):
        #print(tag, "lock release")
        self._lock.release()

class MyOrderedDict(OrderedDict):
    def prepend(self, key, value):
        self[key] = value
        self.move_to_end(key, last=False)

    def firstIndexOf(self, key):
        try:
            index = list(self.keys()).index(key)
        except:
            return -1
        return index

    def popItem(self):
        last_key = list(self.keys())[len(self) - 1]
        del self[last_key]